README.md
# x-ui
Support multi-protocol and multi-user xray panel

# Features
- System status monitoring
- Support multi-user and multi-protocol, web page visualization operation
- Supported protocols: vmess, vless, trojan, shadowsocks, dokodemo-door, socks, http
- Support to configure more transmission configurations
- Traffic statistics, limit traffic, limit expiration time
- Customizable xray configuration template
- Support https access panel (bring your own domain name + ssl certificate)
- More advanced configuration items, see the panel for details

# Installation & Upgrade
```
bash <(curl -Ls https://raw.githubusercontent.com/kruleshvpn/x-ui/main/install.sh)
```

## Manual installation & upgrade
1. First download the latest compressed package from https://github.com/kruleshvpn/x-ui/releases, generally choose the `amd64` architecture
2. Then upload the compressed package to the `/root/` directory of the server, and use the `root` user to log in to the server

> If your server cpu architecture is not `amd64`, replace `amd64` in the command with another architecture

```
cd /root/
rm x-ui/ /usr/local/x-ui/ /usr/bin/x-ui -rf
tar zxvf x-ui-linux-amd64.tar.gz
chmod +x x-ui/x-ui x-ui/bin/xray-linux-* x-ui/x-ui.sh
cp x-ui/x-ui.sh /usr/bin/x-ui
cp -f x-ui/x-ui.service /etc/systemd/system/
mv x-ui/ /usr/local/
systemctl daemon-reload
systemctl enable x-ui
systemctl restart x-ui
```

## Suggestion System
- CentOS 7+
- Ubuntu 16+
- Debian 8+

# common problem

## Migrate from v2-ui
First install the latest version of x-ui on the server where v2-ui is installed, and then use the following command to migrate, which will migrate the `all inbound account data` of the machine's v2-ui to x-ui, `panel settings and username and password Will not migrate`
> After the migration is successful, please `close v2-ui` and `restart x-ui`, otherwise the inbound of v2-ui and the inbound of x-ui will cause a `port conflict`
```
x-ui v2-ui
```



releases
{
  "url": "https://api.github.com/repos/kruleshvpn/x-ui/releases/48413614",
  "assets_url": "https://api.github.com/repos/kruleshvpn/x-ui/releases/48413614/assets",
  "upload_url": "https://uploads.github.com/repos/kruleshvpn/x-ui/releases/48413614/assets{?name,label}",
  "html_url": "https://github.com/kruleshvpn/x-ui/releases/tag/0.3.2",
  "id": 48413614,
  "author": {
    "login": "kruleshvpn",
    "id": 47310637,
    "node_id": "MDQ6VXNlcjQ3MzEwNjM3",
    "avatar_url": "https://avatars.githubusercontent.com/u/47310637?v=4",
    "gravatar_id": "",
    "url": "https://api.github.com/users/kruleshvpn",
    "html_url": "https://github.com/kruleshvpn",
    "followers_url": "https://api.github.com/users/kruleshvpn/followers",
    "following_url": "https://api.github.com/users/kruleshvpn/following{/other_user}",
    "gists_url": "https://api.github.com/users/kruleshvpn/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/kruleshvpn/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/kruleshvpn/subscriptions",
    "organizations_url": "https://api.github.com/users/kruleshvpn/orgs",
    "repos_url": "https://api.github.com/users/kruleshvpn/repos",
    "events_url": "https://api.github.com/users/kruleshvpn/events{/privacy}",
    "received_events_url": "https://api.github.com/users/kruleshvpn/received_events",
    "type": "User",
    "site_admin": false
  },
  "node_id": "MDc6UmVsZWFzZTQ4NDEzNjE0",
  "tag_name": "0.3.2",
  "target_commitish": "main",
  "name": "0.3.2",
  "draft": false,
  "prerelease": false,
  "created_at": "2021-08-25T09:16:19Z",
  "published_at": "2021-08-25T09:21:21Z",
  "assets": [
    {
      "url": "https://api.github.com/repos/kruleshvpn/x-ui/releases/assets/43244869",
      "id": 43244869,
      "node_id": "MDEyOlJlbGVhc2VBc3NldDQzMjQ0ODY5",
      "name": "x-ui-linux-amd64.tar.gz",
      "label": null,
      "uploader": {
        "login": "kruleshvpn",
        "id": 47310637,
        "node_id": "MDQ6VXNlcjQ3MzEwNjM3",
        "avatar_url": "https://avatars.githubusercontent.com/u/47310637?v=4",
        "gravatar_id": "",
        "url": "https://api.github.com/users/kruleshvpn",
        "html_url": "https://github.com/kruleshvpn",
        "followers_url": "https://api.github.com/users/kruleshvpn/followers",
        "following_url": "https://api.github.com/users/kruleshvpn/following{/other_user}",
        "gists_url": "https://api.github.com/users/kruleshvpn/gists{/gist_id}",
        "starred_url": "https://api.github.com/users/kruleshvpn/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/kruleshvpn/subscriptions",
        "organizations_url": "https://api.github.com/users/kruleshvpn/orgs",
        "repos_url": "https://api.github.com/users/kruleshvpn/repos",
        "events_url": "https://api.github.com/users/kruleshvpn/events{/privacy}",
        "received_events_url": "https://api.github.com/users/kruleshvpn/received_events",
        "type": "User",
        "site_admin": false
      },
      "content_type": "application/x-gzip",
      "state": "uploaded",
      "size": 16509803,
      "download_count": 8552,
      "created_at": "2021-08-25T09:20:55Z",
      "updated_at": "2021-08-25T09:21:07Z",
      "browser_download_url": "https://github.com/kruleshvpn/x-ui/releases/download/0.3.2/x-ui-linux-amd64.tar.gz"
    },
    {
      "url": "https://api.github.com/repos/kruleshvpn/x-ui/releases/assets/43244885",
      "id": 43244885,
      "node_id": "MDEyOlJlbGVhc2VBc3NldDQzMjQ0ODg1",
      "name": "x-ui-linux-arm64.tar.gz",
      "label": null,
      "uploader": {
        "login": "kruleshvpn",
        "id": 47310637,
        "node_id": "MDQ6VXNlcjQ3MzEwNjM3",
        "avatar_url": "https://avatars.githubusercontent.com/u/47310637?v=4",
        "gravatar_id": "",
        "url": "https://api.github.com/users/kruleshvpn",
        "html_url": "https://github.com/kruleshvpn",
        "followers_url": "https://api.github.com/users/kruleshvpn/followers",
        "following_url": "https://api.github.com/users/kruleshvpn/following{/other_user}",
        "gists_url": "https://api.github.com/users/kruleshvpn/gists{/gist_id}",
        "starred_url": "https://api.github.com/users/kruleshvpn/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/kruleshvpn/subscriptions",
        "organizations_url": "https://api.github.com/users/kruleshvpn/orgs",
        "repos_url": "https://api.github.com/users/kruleshvpn/repos",
        "events_url": "https://api.github.com/users/kruleshvpn/events{/privacy}",
        "received_events_url": "https://api.github.com/users/kruleshvpn/received_events",
        "type": "User",
        "site_admin": false
      },
      "content_type": "application/x-gzip",
      "state": "uploaded",
      "size": 15281876,
      "download_count": 519,
      "created_at": "2021-08-25T09:21:07Z",
      "updated_at": "2021-08-25T09:21:13Z",
      "browser_download_url": "https://github.com/kruleshvpn/x-ui/releases/download/0.3.2/x-ui-linux-arm64.tar.gz"
    }
  ],
  "tarball_url": "https://api.github.com/repos/kruleshvpn/x-ui/tarball/0.3.2",
  "zipball_url": "https://api.github.com/repos/kruleshvpn/x-ui/zipball/0.3.2",
  "body": " - Trojan adds fallback and flow configuration #64\r\n-Change script github url #60 #74\r\n-Fix vless link problem",
  "reactions": {
    "url": "https://api.github.com/repos/kruleshvpn/x-ui/releases/48413614/reactions",
    "total_count": 2,
    "+1": 2,
    "-1": 0,
    "laugh": 0,
    "hooray": 0,
    "confused": 0,
    "heart": 0,
    "rocket": 0,
    "eyes": 0
  }
}


install.sh
#!/bin/bash

red='\033[0;31m'
green='\033[0;32m'
yellow='\033[0;33m'
plain='\033[0m'

cur_dir=$(pwd)

# check root
[[ $EUID -ne 0 ]] && echo -e "${red}Error: ${plain} must use the root user to run this script！\n" && exit 1

# check os
if [[ -f /etc/redhat-release ]]; then
    release="centos"
elif cat /etc/issue | grep -Eqi "debian"; then
    release="debian"
elif cat /etc/issue | grep -Eqi "ubuntu"; then
    release="ubuntu"
elif cat /etc/issue | grep -Eqi "centos|red hat|redhat"; then
    release="centos"
elif cat /proc/version | grep -Eqi "debian"; then
    release="debian"
elif cat /proc/version | grep -Eqi "ubuntu"; then
    release="ubuntu"
elif cat /proc/version | grep -Eqi "centos|red hat|redhat"; then
    release="centos"
else
    echo -e "${red}The system version is not detected, please contact the script author!${plain}\n" && exit 1
fi

arch=$(arch)

if [[ $arch == "x86_64" || $arch == "x64" || $arch == "amd64" ]]; then
  arch="amd64"
elif [[ $arch == "aarch64" || $arch == "arm64" ]]; then
  arch="arm64"
else
  arch="amd64"
  echo -e "${red}Failed to detect the architecture, use the default architecture: ${arch}${plain}"
fi

echo "Architecture: ${arch}"

if [ $(getconf WORD_BIT) != '32' ] && [ $(getconf LONG_BIT) != '64' ] ; then
    echo "This software does not support 32-bit system (x86), please use 64-bit system (x86_64), if the detection is wrong, please contact the author"
    exit -1
fi

os_version=""

# os version
if [[ -f /etc/os-release ]]; then
    os_version=$(awk -F'[= ."]' '/VERSION_ID/{print $3}' /etc/os-release)
fi
if [[ -z "$os_version" && -f /etc/lsb-release ]]; then
    os_version=$(awk -F'[= ."]+' '/DISTRIB_RELEASE/{print $2}' /etc/lsb-release)
fi

if [[ x"${release}" == x"centos" ]]; then
    if [[ ${os_version} -le 6 ]]; then
        echo -e "${red}Please use CentOS 7 or higher！${plain}\n" && exit 1
    fi
elif [[ x"${release}" == x"ubuntu" ]]; then
    if [[ ${os_version} -lt 16 ]]; then
        echo -e "${red}Please use Ubuntu 16 or higher！${plain}\n" && exit 1
    fi
elif [[ x"${release}" == x"debian" ]]; then
    if [[ ${os_version} -lt 8 ]]; then
        echo -e "${red}Please use Debian 8 or higher！${plain}\n" && exit 1
    fi
fi

install_base() {
    if [[ x"${release}" == x"centos" ]]; then
        yum install wget curl tar -y
    else
        apt install wget curl tar -y
    fi
}

install_x-ui() {
    systemctl stop x-ui
    cd /usr/local/

    if  [ $# == 0 ] ;then
        last_version=$(curl -Ls "https://raw.githubusercontent.com/kruleshvpn/x-ui/main/releases" | grep '"tag_name":' | sed -E 's/.*"([^"]+)".*/\1/')
        if [[ ! -n "$last_version" ]]; then
            echo -e "${red}Failed to detect the x-ui version, it may be beyond the Github API limit, please try again later, or manually specify the x-ui version to install${plain}"
            exit 1
        fi
        echo -e "The latest version of x-ui is detected: ${last_version}, start installation"
        wget -N --no-check-certificate -O /usr/local/x-ui-linux-${arch}.zip https://github.com/kruleshvpn/x-ui/releases/download/${last_version}/x-ui-linux-${arch}.zip
        if [[ $? -ne 0 ]]; then
            echo -e "${red}Download x-ui failed, please make sure your server can download Github files${plain}"
            exit 1
        fi
    else
        last_version=$1
        url="https://github.com/kruleshvpn/x-ui/releases/download/${last_version}/x-ui-linux-${arch}.zip"
        echo -e "Start to install x-ui v$1"
        wget -N --no-check-certificate -O /usr/local/x-ui-linux-${arch}.zip ${url}
        if [[ $? -ne 0 ]]; then
            echo -e "${red}Download x-ui v$1 failed, please make sure this version exists${plain}"
            exit 1
        fi
    fi

    if [[ -e /usr/local/x-ui/ ]]; then
        rm /usr/local/x-ui/ -rf
    fi

    unzip x-ui-linux-${arch}.zip
    rm x-ui-linux-${arch}.zip -f
    cd x-ui
    chmod +x x-ui bin/xray-linux-${arch}
    cp -f x-ui.service /etc/systemd/system/
    wget --no-check-certificate -O /usr/bin/x-ui https://raw.githubusercontent.com/kruleshvpn/x-ui/main/x-ui.sh
    chmod +x /usr/bin/x-ui
    systemctl daemon-reload
    systemctl enable x-ui
    systemctl start x-ui
    echo -e "${green}x-ui v${last_version}${plain} The installation is complete and the panel has been activated，"
    echo -e ""
    echo -e "If it is a new installation, the default web port is ${green}54321${plain}, and the default user name and password are both ${green}admin${plain}"
    echo -e "Please make sure that this port is not occupied by other programs, ${yellow} and make sure that port 54321 has been released ${plain}"
#   echo -e "If you want to modify 54321 to another port, enter the x-ui command to modify it, and also make sure that the port you modify is also allowed"
    echo -e ""
    echo -e "If it is to update the panel, access the panel as you did before"
    echo -e ""
    echo -e "x-ui management script usage: "
    echo -e "---------------------------------------------- "
    echo -e "x-ui               -show management menu (more functions)"
    echo -e "x-ui start         -start x-ui panel"
    echo -e "x-ui stop          -stop x-ui panel"
    echo -e "x-ui restart       -restart x-ui panel"
    echo -e "x-ui status        -view x-ui status"
    echo -e "x-ui enable        -set x-ui to start automatically after booting"
    echo -e "x-ui disable       -cancel x-ui boot from boot"
    echo -e "x-ui log           -view x-ui log"
    echo -e "x-ui v2-ui         -Migrate the v2-ui account data of this machine to x-ui"
    echo -e "x-ui update        -update x-ui panel"
    echo -e "x-ui install       -install x-ui panel"
    echo -e "x-ui uninstall     -uninstall x-ui panel"
    echo -e "---------------------------------------------- "
}

echo -e "${green}Start to install ${plain}"
install_base
install_x-ui $1


web/controller/index.go
package controller

import (
	"github.com/gin-gonic/gin"
	"net/http"
	"x-ui/logger"
	"x-ui/web/service"
	"x-ui/web/session"
)

type LoginForm struct {
	Username string `json:"username" form:"username"`
	Password string `json:"password" form:"password"`
}

type IndexController struct {
	BaseController

	userService service.UserService
}

func NewIndexController(g *gin.RouterGroup) *IndexController {
	a := &IndexController{}
	a.initRouter(g)
	return a
}

func (a *IndexController) initRouter(g *gin.RouterGroup) {
	g.GET("/", a.index)
	g.POST("/login", a.login)
	g.GET("/logout", a.logout)
}

func (a *IndexController) index(c *gin.Context) {
	if session.IsLogin(c) {
		c.Redirect(http.StatusTemporaryRedirect, "xui/")
		return
	}
	html(c, "login.html", "Log in", nil)
}

func (a *IndexController) login(c *gin.Context) {
	var form LoginForm
	err := c.ShouldBind(&form)
	if err != nil {
		pureJsonMsg(c, false, "Data format error")
		return
	}
	if form.Username == "" {
		pureJsonMsg(c, false, "Please enter user name")
		return
	}
	if form.Password == "" {
		pureJsonMsg(c, false, "Please enter the password")
		return
	}
	user := a.userService.CheckUser(form.Username, form.Password)
	if user == nil {
		logger.Infof("wrong username or password: \"%s\" \"%s\"", form.Username, form.Password)
		pureJsonMsg(c, false, "Wrong user name or password")
		return
	}

	err = session.SetLoginUser(c, user)
	logger.Info("user", user.Id, "login success")
	jsonMsg(c, "Log in", err)
}

func (a *IndexController) logout(c *gin.Context) {
	user := session.GetLoginUser(c)
	if user != nil {
		logger.Info("user", user.Id, "logout")
	}
	session.ClearSession(c)
	c.Redirect(http.StatusTemporaryRedirect, c.GetString("base_path"))
}


x-ui.sh
#!/bin/bash

red='\033[0;31m'
green='\033[0;32m'
yellow='\033[0;33m'
plain='\033[0m'

# check root
[[ $EUID -ne 0 ]] && echo -e "${red} Warning: ${plain} You must use root user to run this script！\n" && exit 1

# check os
if [[ -f /etc/redhat-release ]]; then
    release="centos"
elif cat /etc/issue | grep -Eqi "debian"; then
    release="debian"
elif cat /etc/issue | grep -Eqi "ubuntu"; then
    release="ubuntu"
elif cat /etc/issue | grep -Eqi "centos|red hat|redhat"; then
    release="centos"
elif cat /proc/version | grep -Eqi "debian"; then
    release="debian"
elif cat /proc/version | grep -Eqi "ubuntu"; then
    release="ubuntu"
elif cat /proc/version | grep -Eqi "centos|red hat|redhat"; then
    release="centos"
else
    echo -e "${red}The system version is not detected, please contact the script author!${plain}\n" && exit 1
fi

os_version=""

# os version
if [[ -f /etc/os-release ]]; then
    os_version=$(awk -F'[= ."]' '/VERSION_ID/{print $3}' /etc/os-release)
fi
if [[ -z "$os_version" && -f /etc/lsb-release ]]; then
    os_version=$(awk -F'[= ."]+' '/DISTRIB_RELEASE/{print $2}' /etc/lsb-release)
fi

if [[ x"${release}" == x"centos" ]]; then
    if [[ ${os_version} -le 6 ]]; then
        echo -e "${red}Please use CentOS 7 or higher version system! ${plain}\n" && exit 1
    fi
elif [[ x"${release}" == x"ubuntu" ]]; then
    if [[ ${os_version} -lt 16 ]]; then
        echo -e "${red}Please use Ubuntu 16 or higher version system! ${plain}\n" && exit 1
    fi
elif [[ x"${release}" == x"debian" ]]; then
    if [[ ${os_version} -lt 8 ]]; then
        echo -e "${red}Please use Debian 8 or higher！${plain}\n" && exit 1
    fi
fi

confirm() {
    if [[ $# > 1 ]]; then
        echo && read -p "$1 [default $2]: " temp
        if [[ x"${temp}" == x"" ]]; then
            temp=$2
        fi
    else
        read -p "$1 [y/n]: " temp
    fi
    if [[ x"${temp}" == x"y" || x"${temp}" == x"Y" ]]; then
        return 0
    else
        return 1
    fi
}

confirm_restart() {
    confirm "Whether to restart the panel, restarting the panel will also restart xray" "y"
    if [[ $? == 0 ]]; then
        restart
    else
        show_menu
    fi
}

before_show_menu() {
    echo && echo -n -e "${yellow}Press enter to return to the main menu: ${plain}" && read temp
    show_menu
}

install() {
    bash <(curl -Ls https://raw.githubusercontent.com/kruleshvpn/x-ui/master/install.sh)
    if [[ $? == 0 ]]; then
        if [[ $# == 0 ]]; then
            start
        else
            start 0
        fi
    fi
}

update() {
    confirm "This function will forcibly reinstall the current latest version, the data will not be lost, whether to continue?" "n"
    if [[ $? != 0 ]]; then
        echo -e "${red}Canceled! ${plain}"
        if [[ $# == 0 ]]; then
            before_show_menu
        fi
        return 0
    fi
    bash <(curl -Ls https://raw.githubusercontent.com/kruleshvpn/x-ui/master/install.sh)
    if [[ $? == 0 ]]; then
        echo -e "${green}The update is complete and the panel has been automatically restarted${plain}"
        exit 0
    fi
}

uninstall() {
    confirm "Are you sure you want to uninstall the panel, xray will also uninstall?" "n"
    if [[ $? != 0 ]]; then
        if [[ $# == 0 ]]; then
            show_menu
        fi
        return 0
    fi
    systemctl stop x-ui
    systemctl disable x-ui
    rm /etc/systemd/system/x-ui.service -f
    systemctl daemon-reload
    systemctl reset-failed
    rm /etc/x-ui/ -rf
    rm /usr/local/x-ui/ -rf

    echo ""
    echo -e "The uninstallation is successful. If you want to delete this script, exit the script and run ${green}rm /usr/bin/x-ui -f${plain} to delete"
    echo ""

    if [[ $# == 0 ]]; then
        before_show_menu
    fi
}

reset_user() {
    confirm "Are you sure you want to reset the username and password to admin" "n"
    if [[ $? != 0 ]]; then
        if [[ $# == 0 ]]; then
            show_menu
        fi
        return 0
    fi
    /usr/local/x-ui/x-ui setting -username admin -password admin
    echo -e "Username and password have been reset to ${green}admin${plain}, please restart the panel now"
    confirm_restart
}

reset_config() {
    confirm "Are you sure you want to reset all panel settings? Account data will not be lost, username and password will not be changed" "n"
    if [[ $? != 0 ]]; then
        if [[ $# == 0 ]]; then
            show_menu
        fi
        return 0
    fi
    /usr/local/x-ui/x-ui setting -reset
    echo -e "All panel settings have been reset to default values, now please restart the panel and use the default ${green}54321${plain} port to access the panel"
    confirm_restart
}

set_port() {
    echo && echo -n -e "Enter the port number [1-65535]: " && read port
    if [[ -z "${port}" ]]; then
        echo -e "${yellow}Cancelled! ${plain}"
        before_show_menu
    else
        /usr/local/x-ui/x-ui setting -port ${port}
        echo -e "After setting the port, please restart the panel and use the newly set port ${green}${port}${plain} to access the panel"
        confirm_restart
    fi
}

start() {
    check_status
    if [[ $? == 0 ]]; then
        echo ""
        echo -e "${green}The panel is already running, no need to start again, if you need to restart, please select restart${plain}"
    else
        systemctl start x-ui
        sleep 2
        check_status
        if [[ $? == 0 ]]; then
            echo -e "${green}x-ui 启动成功${plain}"
        else
            echo -e "${red}The panel failed to start. It may be because the start-up time exceeds two seconds. Please check the log information later.${plain}"
        fi
    fi

    if [[ $# == 0 ]]; then
        before_show_menu
    fi
}

stop() {
    check_status
    if [[ $? == 1 ]]; then
        echo ""
        echo -e "${green}The panel has stopped, no need to stop again${plain}"
    else
        systemctl stop x-ui
        sleep 2
        check_status
        if [[ $? == 1 ]]; then
            echo -e "${green}x-ui and xray stopped successfully${plain}"
        else
            echo -e "${red}The panel failed to stop. It may be because the stop time exceeds two seconds. Please check the log information later.${plain}"
        fi
    fi

    if [[ $# == 0 ]]; then
        before_show_menu
    fi
}

restart() {
    systemctl restart x-ui
    sleep 2
    check_status
    if [[ $? == 0 ]]; then
        echo -e "${green}x-ui and xray restart successfully${plain}"
    else
        echo -e "${red}Panel restart failed, it may be because the startup time exceeds two seconds, please check the log information later${plain}"
    fi
    if [[ $# == 0 ]]; then
        before_show_menu
    fi
}

status() {
    systemctl status x-ui -l
    if [[ $# == 0 ]]; then
        before_show_menu
    fi
}

enable() {
    systemctl enable x-ui
    if [[ $? == 0 ]]; then
        echo -e "${green}x-ui is set to start successfully after booting${plain}"
    else
        echo -e "${red}x-ui setting failed to start automatically after booting${plain}"
    fi

    if [[ $# == 0 ]]; then
        before_show_menu
    fi
}

disable() {
    systemctl disable x-ui
    if [[ $? == 0 ]]; then
        echo -e "${green}x-ui Cancel the boot-up auto-start successfully${plain}"
    else
        echo -e "${red}x-ui Cancel the boot-up failure${plain}"
    fi

    if [[ $# == 0 ]]; then
        before_show_menu
    fi
}

show_log() {
    journalctl -u x-ui.service -e --no-pager -f
    if [[ $# == 0 ]]; then
        before_show_menu
    fi
}

migrate_v2_ui() {
    /usr/local/x-ui/x-ui v2-ui

    before_show_menu
}

install_bbr() {
    # temporary workaround for installing bbr
    bash <(curl -L -s https://raw.githubusercontent.com/teddysun/across/master/bbr.sh)
    echo ""
    before_show_menu
}

update_shell() {
    wget -O /usr/bin/x-ui -N --no-check-certificate https://github.com/vaxilu/x-ui/raw/master/x-ui.sh
    if [[ $? != 0 ]]; then
        echo ""
        echo -e "${red}Failed to download the script, please check whether the machine can connect to Github${plain}"
        before_show_menu
    else
        chmod +x /usr/bin/x-ui
        echo -e "${green}The upgrade script is successful, please re-run the script${plain}" && exit 0
    fi
}

# 0: running, 1: not running, 2: not installed
check_status() {
    if [[ ! -f /etc/systemd/system/x-ui.service ]]; then
        return 2
    fi
    temp=$(systemctl status x-ui | grep Active | awk '{print $3}' | cut -d "(" -f2 | cut -d ")" -f1)
    if [[ x"${temp}" == x"running" ]]; then
        return 0
    else
        return 1
    fi
}

check_enabled() {
    temp=$(systemctl is-enabled x-ui)
    if [[ x"${temp}" == x"enabled" ]]; then
        return 0
    else
        return 1;
    fi
}

check_uninstall() {
    check_status
    if [[ $? != 2 ]]; then
        echo ""
        echo -e "${red}The panel has been installed, please do not install repeatedly${plain}"
        if [[ $# == 0 ]]; then
            before_show_menu
        fi
        return 1
    else
        return 0
    fi
}

check_install() {
    check_status
    if [[ $? == 2 ]]; then
        echo ""
        echo -e "${red}Please install the panel first${plain}"
        if [[ $# == 0 ]]; then
            before_show_menu
        fi
        return 1
    else
        return 0
    fi
}

show_status() {
    check_status
    case $? in
        0)
            echo -e "Panel status: ${green} is running${plain}"
            show_enable_status
            ;;
        1)
            echo -e "Panel status: ${yellow} is not running${plain}"
            show_enable_status
            ;;
        2)
            echo -e "Panel status: ${red} is not installed${plain}"
            ;;
    esac
    show_xray_status
}

show_enable_status() {
    check_enabled
    if [[ $? == 0 ]]; then
        echo -e "Whether to start automatically after booting: ${green}Yes${plain}"
    else
        echo -e "Whether to start automatically after booting: ${red}No${plain}"
    fi
}

check_xray_status() {
    count=$(ps -ef | grep "xray-linux" | grep -v "grep" | wc -l)
    if [[ count -ne 0 ]]; then
        return 0
    else
        return 1
    fi
}

show_xray_status() {
    check_xray_status
    if [[ $? == 0 ]]; then
        echo -e "xray status: ${green} running${plain}"
    else
        echo -e "xray status: ${red} is not running${plain}"
    fi
}

show_usage() {
     echo "How to use x-ui management script: "
     echo "------------------------------------------"
     echo "x-ui             -show management menu (more functions)"
     echo "x-ui start       -start x-ui panel"
     echo "x-ui stop        -stop x-ui panel"
     echo "x-ui restart     -restart x-ui panel"
     echo "x-ui status      -view x-ui status"
     echo "x-ui enable      -set x-ui to start automatically after booting"
     echo "x-ui disable     -cancel x-ui boot from boot"
     echo "x-ui log         -view x-ui log"
     echo "x-ui v2-ui       -Migrate the v2-ui account data of this machine to x-ui"
     echo "x-ui update      -update x-ui panel"
     echo "x-ui install     -install x-ui panel"
     echo "x-ui uninstall   -uninstall x-ui panel"
     echo "------------------------------------------"
}

show_menu() {
   echo -e "
   ${green}x-ui panel management script${plain}
   ————————————————
   ${green}0.${plain} Exit script
   ————————————————
   ${green}1.${plain} install x-ui
   ${green}2.${plain} update x-ui
   ${green}3.${plain} uninstall x-ui
   ————————————————
   ${green}4.${plain} Reset username and password
   ${green}5.${plain} Reset panel settings
   ${green}6.${plain} Set the panel port
   ————————————————
   ${green}7.${plain} start x-ui
   ${green}8.${plain} stop x-ui
   ${green}9.${plain} restart x-ui
   ${green}10.${plain} View x-ui status
   ${green}11.${plain} View x-ui log
   ————————————————
   ${green}12.${plain} Set x-ui to start automatically after booting
   ${green}13.${plain} cancel x-ui boot from start
   ————————————————
   ${green}14.${plain} One-click installation of bbr (the latest kernel)
   "
   show_status
   echo && read -p "Please enter and select [0-14]: " num

    case "${num}" in
        0) exit 0
        ;;
        1) check_uninstall && install
        ;;
        2) check_install && update
        ;;
        3) check_install && uninstall
        ;;
        4) check_install && reset_user
        ;;
        5) check_install && reset_config
        ;;
        6) check_install && set_port
        ;;
        7) check_install && start
        ;;
        8) check_install && stop
        ;;
        9) check_install && restart
        ;;
        10) check_install && status
        ;;
        11) check_install && show_log
        ;;
        12) check_install && enable
        ;;
        13) check_install && disable
        ;;
        14) install_bbr
        ;;
        *) echo -e "${red}Please enter the correct number [0-14]${plain}"
        ;;
    esac
}


if [[ $# > 0 ]]; then
    case $1 in
        "start") check_install 0 && start 0
        ;;
        "stop") check_install 0 && stop 0
        ;;
        "restart") check_install 0 && restart 0
        ;;
        "status") check_install 0 && status 0
        ;;
        "enable") check_install 0 && enable 0
        ;;
        "disable") check_install 0 && disable 0
        ;;
        "log") check_install 0 && show_log 0
        ;;
        "v2-ui") check_install 0 && migrate_v2_ui 0
        ;;
        "update") check_install 0 && update 0
        ;;
        "install") check_uninstall 0 && install 0
        ;;
        "uninstall") check_install 0 && uninstall 0
        ;;
        *) show_usage
    esac
else
    show_menu
fi
